syntax = "proto3";

service KVStore {
  rpc Put(PutRequest) returns (PutResponse) {}
  rpc Get(GetRequest) returns (GetResponse) {}
  rpc CPut(CPutRequest) returns (CPutResponse) {}
  rpc Delete(DeleteRequest) returns (DeleteResponse) {}
  rpc PPut(PutRequest) returns (PutResponse) {}
  rpc PDelete(DeleteRequest) returns (DeleteResponse) {}

  rpc TabletSync(TabletSyncRequest) returns (stream KeyValue) {}
  rpc HealthCheck(HealthRequest) returns (HealthResponse) {}
  rpc PrimaryUpdate(PrimaryRequest) returns (PrimaryResponse) {}

  rpc AdminList(ListRequest) returns (ListResponse) {}
  rpc AdminStartup(StartupRequest) returns (StartupResponse) {}
  rpc AdminShutdown(ShutdownRequest) returns (ShutdownResponse) {}
}

message PutRequest {
  string row = 1;
  string col = 2;
  bytes value = 3; 
}

message GetRequest {
  string row = 1;
  string col = 2;
}

message CPutRequest {
  string row = 1;
  string col = 2;
  bytes v1 = 3;  
  bytes v2 = 5; 
}

message DeleteRequest {
  string row = 1;
  string col = 2;
}

message PutResponse {
  uint32 bytes_written = 1;
}

message GetResponse {
  KeyValue data = 1;
}

message CPutResponse {
  uint32 bytes_written = 1;
}

message DeleteResponse {
  uint32 keys_deleted = 1;
}

message TabletSyncRequest {}

message HealthRequest {}

message PrimaryRequest {
  string addr = 1;  
}

message ListRequest {}
message StartupRequest {}
message ShutdownRequest {}

message KeyValue {
  string row = 1;
  string col = 2;
  bytes value = 3;  
}

message Tablet {
  message Row {
    map<string, bytes> cols = 1;
  }

  map<string, Row> rows = 1;
}
message HealthResponse {}
message PrimaryResponse {}

message ListResponse {
  message KVPair {
    string row = 1;
    string col = 2;
  }
  repeated KVPair keys = 1;
}
message StartupResponse {}
message ShutdownResponse {}
